### PR TITLE
Refactor Otoken minting

### PR DESCRIPTION
**Pull request by haythem96**
# Refactor Otoken minting

## High Level Description

Remove external calls to `AddressBook`.
To apply this change, need to redeploy Factory and new Otoken implementation.
With this change, we need to deploy new Otoken implementation whenever we upgrade Controller.

### Code

- [x] Unit test 100% coverage
- [x] Does your code follow the naming and code documentation guidelines?

### Documentation

- [ ] Is your code up to date with the spec? 
- [ ] Have you added your tests to the testing doc?

### CODE REVIEW

**Comment by haythem96** [comment]
Screen Shot 2021-01-11 at 7 51 42 AM

**Review by antoncoding — APPROVED** [review]
From my understanding, this refactor works perfectly because the fact that **we will never update our controller address in `Addressbook`**, which I think make sense but want every body to give it a thought as well.

**Comment by haythem96** [comment]
> From my understanding, this refactor works perfectly because the fact that **we will never update our controller address in `Addressbook`**, which I think make sense but want every body to give it a thought as well.

Ah yeah that's true, I forgot that we don't need to change the Controller address as the Proxy will always forward to the right implementation.

**Comment by antoncoding** [comment]
From our discussion:
Better keep the addressbook in the oToken init function to avoid updating factory contract. We can store the controller address locally during initialization so we can reduce that 1 external call every time mint and burn is called.

**Comment by haythem96** [comment]
> From our discussion:
> Better keep the addressbook in the oToken init function to avoid updating factory contract. We can store the controller address locally during initialization so we can reduce that 1 external call every time mint and burn is called.

Yeah agree, I updated the code.

File: contracts/Otoken.sol [hunk 1]
```diff
@@ -16,8 +16,8 @@ import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
 contract Otoken is ERC20Initializable {
     using SafeMath for uint256;
 
-    /// @notice address of the AddressBook module
-    address public addressBook;
+    /// @notice address of the Controller module
+    address public controller;
```
Comment (aparnakr) [inline contracts/Otoken.sol:20]: if we upgrade the controller, is there a way to set this again?

**Review by aparnakr — COMMENTED** [review]
It looks good! Left a question which we should figure out first!

File: contracts/Otoken.sol (hunk 1 above)
Comment (haythem96) [reply contracts/Otoken.sol:20]: Yeah, when we upgrade the controller, the controller address does not change, it is always the same proxy address, the only address that change is the implementation address which is all handled by the Proxy mechanism.

**Review by antoncoding — APPROVED** [review]

File: contracts/Otoken.sol (hunk 1 above)
Comment (aparnakr) [reply contracts/Otoken.sol:20]: can we test upgrading the controller with this?

**Review by aparnakr — APPROVED** [review]

**Pull request merged**